src/bin/languageserver/mod.rs

```diff
@@ -1158,18 +1163,78 @@ impl LanguageServer for SolangServer {
     async fn did_open(&self, params: DidOpenTextDocumentParams) {
         let uri = params.text_document.uri;
 
+        if let Ok(path) = uri.to_file_path() {
+            self.text_buffers.lock().await.insert(path, params.text_document.text);
+        }
+
         self.parse_file(uri).await;
     }
 
     async fn did_change(&self, params: DidChangeTextDocumentParams) {
         let uri = params.text_document.uri;
 
+        if let Ok(path) = uri.to_file_path() {
+            if let Some(text_buf) = self.text_buffers.lock().await.get_mut(&path) {
+                for content_change in params.content_changes {
+                    *text_buf = if let Some(range) = content_change.range {
+                        let start_line = range.start.line as usize;
+                        let start_col = range.start.character as usize;
+                        let end_line = range.end.line as usize;
+                        let end_col = range.end.character as usize;
+                        
+                        // directly add the changes to the buffer when changes are present at the end of the file
+                        if start_line == text_buf.lines().count() {
+                            text_buf.push_str(&content_change.text);
+                            return;
+                        }
+
+                        let mut new = String::new();
+                        for (i, line) in text_buf.lines().enumerate() {
+                            if i < start_line {
+                                new.push_str(&line);
+                                new.push('\n');
+                                continue;
+                            }
+
+                            if i > end_line {
+                                new.push_str(&line);
+                                new.push('\n');
+                                continue;
+                            }
+
+                            if i == start_line {
+                                new.push_str(&line[..start_col]);
+                                new.push_str(&content_change.text);
+                            }
+
+                            if i == end_line {
+                                new.push_str(&line[end_col..]);
+                                new.push('\n');
+                            }
+                        }
+                        new
+                    } else {
+                        // when no range is provided, entire file is sent in the request
+                        content_change.text
+                    };
+                }
+            }
+        }
+
```

Comment:
This code could really do with some tests.

Reply:
I have added the tests in the same file `src/bin/languageserver/mod.rs`. 
Do I need to move them to a different location? In that case I will have to make the function that I am testing `public`. Not sure if that's okay.

Reply:
That's fine. If the file gets too big it might be time to split it up.